design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,74 @@
+# Ceph COSI Driver Support
+
+## Targeted for v1.12
+
+## Background
+
+Container Object Storage Interface (COSI) is a set of specifications for container orchestration frameworks to manage object storage. Even though there is no standard procotol defined for Object Store, it has flexibility to add support for all. The COSI spec abstracts common storage features such as create/delete buckets, grant/revoke access to buckets, attach/detach buckets, and more. It is currently at the alpha release. COSI is a new project and is not yet fully integrated by Kubernetes.
+More details about COSI can be found [here](https://kubernetes.io/blog/2022/09/02/cosi-kubernetes-object-storage-management/)
+It is projected that COSI will be the only supported object storage driver in the near feature. In-tree drivers such as Ceph RGW will be replaced with their respective COSI drivers.
+
+## Current Status of Ceph COSI Driver
+
+The [Ceph COSI driver](https://github.com/ceph/ceph-cosi) is currently in the pre-alpha status. It is currently being tested with latest COSI Spec and APIs with images:
+
+- cosi-controller : gcr.io/k8s-staging-sig-storage/objectstorage-controller:v20230130-v0.1.1-12-geafd6fa
+- cosi-sidecar : gcr.io/k8s-staging-sig-storage/objectstorage-sidecar/objectstorage-sidecar:v20230130-v0.1.0-24-gc0cf995
+
+The Ceph COSI driver supports RGW with ceph storage as backend with s3 protocol. Later on, it will be extended to support other protocols such as Swift and different backends for RGW via Zipper.
+
+## COSI Driver Deployment
+
+The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
```

Comment:
> The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace. 

It really has to be in the default namespace? If it had to be a specific a specific namespace, I'm surprised it's not `kube-system`.

> The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). 

What state does the driver have? Why does it need to be a statefulset? Would it be backed by an rbd volume? 

> The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:

Let's expand on this more. Based on our conversation, sounds like we would put it in the same namespace as the CephObjectStore CR, unless we decide there is a need to put it with the operator.

Reply:
> > The [COSI controller](https://github.com/kubernetes-sigs/container-object-storage-interface-controller) is deployed as container in the default namespace.
> 
> It really has to be in the default namespace? If it had to be a specific a specific namespace, I'm surprised it's not `kube-system`.

Atleast they resources are defined like setting up in default namespace https://github.com/kubernetes-sigs/container-object-storage-interface-controller/tree/master/resources
> 
> > The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar).
> 
> What state does the driver have? Why does it need to be a statefulset? Would it be backed by an rbd volume?
> 

Currently, it is stateless, it does not have data stored internally atm, not sure whether we need storage. 

> > The Ceph COSI driver can be deployed in any namespace not along with the COSI controller. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
> 
> Let's expand on this more. Based on our conversation, sounds like we would put it in the same namespace as the CephObjectStore CR, unless we decide there is a need to put it with the operator.

Sure

Reply:
It seems as though we should...
1. Figure out if the COSI controller needs installed in the default namespace. It seems to me that the namespace should be able to be arbitrary unless there is some unnecessary coding assumption in the COSI controller. This doc should only need to be detailed enough to explain that the COSI controller is a necessary prerequisite with a link to COSI docs about the controller itself.
2. Change the Ceph-COSI driver to be a deployment since we don't need a statefulset currently.

-----

I also see some implicit decision-making happening here that I would prefer to have a record of. Technical designs shouldn't leave any decision as implicit; we want to have a reason for each one, and the easiest way to do that is to actually discuss each option that is available to us.

I also think we include in the doc a pro/con list of the different namespaces to install the driver in.
1. The rook system namespace such that it operates for multiple ceph clusters
2. The cluster namespace such that the driver is run once for each ceph cluster

We should also include a pro/con list of how many COSI drivers should be run:
1. Just one in the whole k8s system
2. One for every rook operator
4. One for every CephCluster
5. One for every CephObjectStore

These decisions aren't necessarily just for Rook. These decisions also affect the design of the Ceph COSI controller itself. 

-----

As I think about it more for myself, I think the best decision would be to install one Ceph COSI driver (with the option of multiple replicas) in the whole system. Rook's CephObjectStore controller should then create the necessary resources to allow the COSI driver to allow bucket provisioning for said CephObjectStore.

This decision does not assume that the decisions made for Ceph-CSI are the best or correct ones.

For the pros, I believe that this more closely matches the design paradigms of Kubernetes and other CNCF projects:
- Similarly to the k8s API server, there is only one, and it can scale to meet demand
- Because there is only one, cloud install environments (where CPU and memory demands cost ever more money) don't need to worry about excess resource requirements due to needing more instances than necessary
- The Ceph-COSI driver itself is an independent component from Rook
  - This allows other components to integrate with Ceph-COSI if desired without needing Rook 
  - This means that Rook won't affect the ability of other components to integrate with Ceph-COSI
  - Rook can use Ceph-COSI's defined APIs (whether CRDs or via Secrets/ConfigMaps) to integrate itself as desired
- Rook's codebase does not need to codify or document how to install or configure Ceph-COSI. The responsibility for that is moved to the ceph/ceph-cosi project itself which reduces duplication of effort and reduces the possibility of copy-paste errors between the projects which we sometimes see with Ceph-CSI. It allows Ceph-CSI to make necessary changes (especially to its deployment) more quickly without need for changes to the Rook codebase.

The downsides are that:
- Rook doesn't control Ceph-COSI, and so it might be modified or removed such that Rook's integration wouldn't work
  - Having a well-defined and backwards-compatible Ceph-COSI API will help with this
  - This is in some way the nature of everything in the CNCF ecosystem. We could recommend that users put a finalizer on the Ceph-COSI deployment when Rook is installed stating Rook's intended and ongoing usage of it, which helps mitigate this issue (and is one reason why finalizers exist)

-----

On that note, we should have a design review once the alpha COSI controller is done. I have seen our group often make the mistake of assuming that an initially-working piece of software is a good foundation to build off of for the future, when that is most definitely not the case in any of my experience.

In particular, I want to push to make sure the COSI driver is using controller-runtime and is making full use of structured logging. I would also suggest having a versioned, backwards/forwards-compatible API for Ceph-COSI's configs that can be used by Rook.

@thotz, when Ceph-COSI alpha is done or mostly done, let's organize a design review with the important stakeholders. That would be me and Travis from Rook; Ali who also bridges the RGW-Rook gap; and this could also include Matt B. and/or Casey B. on the RGW side.